src/core/config/npo/tracker.rs

```diff
@@ -92,80 +92,119 @@ impl<'a> FieldName<'a> {
+    // Caches resolved chunks for the specific type
+    // with is_list info since the result is different depending on this flag
+    cache: HashMap<(TypeName<'a>, bool), Chunk<Chunk<Name<'a>>>>,
 }
 
 impl<'a> PathTracker<'a> {
     pub fn new(config: &'a Config) -> PathTracker<'a> {
         PathTracker { config, cache: Default::default() }
     }
 
-    #[allow(clippy::too_many_arguments)]
     fn iter(
         &mut self,
-        path: Chunk<FieldName<'a>>,
+        parent_name: Option<Name<'a>>,
         type_name: TypeName<'a>,
         is_list: bool,
-        visited: HashSet<(TypeName<'a>, FieldName<'a>)>,
-    ) -> Chunk<Chunk<FieldName<'a>>> {
-        if let Some(chunks) = self.cache.get(&(type_name, is_list)) {
-            return chunks.clone();
-        }
+    ) -> Chunk<Chunk<Name<'a>>> {
+        let chunks = if let Some(chunks) = self.cache.get(&(type_name, is_list)) {
+            chunks.clone()
+        } else {
+            // set empty value in the cache to prevent infinity recursion
+            self.cache.insert((type_name, is_list), Chunk::default());
+
+            let mut chunks = Chunk::default();
+            if let Some(type_of) = self.config.find_type(type_name.as_str()) {
+                for (name, field) in type_of.fields.iter() {
+                    let field_name = Name::Field(FieldName::new(name));
 
-        let mut chunks = Chunk::default();
-        if let Some(type_of) = self.config.find_type(type_name.as_str()) {
-            for (name, field) in type_of.fields.iter() {
-                let field_name = FieldName::new(name);
-                let path = path.clone().append(field_name);
-                if !visited.contains(&(type_name, field_name)) {
                     if is_list && field.has_resolver() && !field.has_batched_resolver() {
-                        chunks = chunks.append(path.clone());
+                        chunks = chunks.append(Chunk::new(field_name));
                     } else {
-                        let mut visited = visited.clone();
-                        visited.insert((type_name, field_name));
                         let is_list = is_list | field.type_of.is_list();
                         chunks = chunks.concat(self.iter(
-                            path,
+                            Some(field_name),
                             TypeName::new(field.type_of.name()),
                             is_list,
-                            visited,
                         ))
                     }
                 }
             }
-        }
 
-        self.cache.insert((type_name, is_list), chunks.clone());
-        chunks
+            self.cache.insert((type_name, is_list), chunks.clone());
+
+            chunks
+        };
+
+        // chunks contains only paths from the current type.
+        // Prepend every subpath with parent path
+        if let Some(path) = parent_name {
+            chunks.transform(move |chunk| Chunk::new(path).concat(chunk))
```

Comment:
this doesn't work since the `'static` constraint on the closure in https://github.com/tailcallhq/tailcall-chunk/blob/9fb3a24c5340c3d340491de6ae35b6ddbac60ece/src/chunk.rs#L220 limiting usage of Borrowed data almost completely. 

I'm not sure how we can resolve it except using owned data in chunk and copy strings

Reply:
fixed by using `as_vec`